rocAL/rocAL_pybind/amd/rocal/plugin/generic.py

```diff
@@ -38,50 +38,63 @@ def __init__(self, pipeline, tensor_layout = types.NCHW, reverse_channels = Fals
         self.batch_size = pipeline._batch_size
         if self.loader._name is None:
             self.loader._name = self.loader._reader
-        self.labels_size = ((self.bs*self.loader._numOfClasses) if (self.loader._oneHotEncoding == True) else self.bs)
-        self.out = self.dimensions = self.dtype = None
-        self.len = b.getRemainingImages(self.loader._handle)//self.bs # iteration length
+        self.labels_size = ((self.batch_size*self.loader._numOfClasses) if (self.loader._oneHotEncoding == True) else self.batch_size)
+        self.output_list = self.dimensions = self.dtype = None
+        self.len = b.getRemainingImages(self.loader._handle)//self.batch_size # iteration length
 
     def next(self):
         return self.__next__()
 
     def __next__(self):
         
-        if(b.isEmpty(self.loader._handle)):
+        if self.loader.rocalRun() != 0:
             raise StopIteration
         else:
             self.output_tensor_list = self.loader.getOutputTensors()
 
-        if self.out is None:
-            self.dimensions = self.output_tensor_list[0].dimensions()
-            if self.device == "cpu":
-                self.dtype = self.output_tensor_list[0].dtype()
-                self.out = np.empty((self.dimensions[0], self.dimensions[1], self.dimensions[2], self.dimensions[3],), dtype = self.dtype)
-                self.labels = np.empty(self.labels_size, dtype = self.dtype)
-            else:
-                self.dtype = self.output_tensor_list[0].dtype()
-                with cp.cuda.Device(device = self.device_id):
-                    self.out = cp.empty((self.dimensions[0], self.dimensions[1], self.dimensions[2], self.dimensions[3],), dtype = self.dtype)
-                    self.labels = cp.empty(self.labels_size, dtype = self.dtype)
+        if self.output_list is None:
+            self.output_list = []
+            for i in range(len(self.output_tensor_list)):
+                self.dimensions = self.output_tensor_list[i].dimensions()
+                if self.device == "cpu":
+                    self.dtype = self.output_tensor_list[i].dtype()
+                    self.out = np.empty(self.dimensions, dtype = self.dtype)
+                    self.labels = np.empty(self.labels_size, dtype = self.dtype)
```

Comment:
Do you think it is better to use np.zeros instead of np.empty?

Reply:
From numpy docs,
np.empty, unlike np.zeros, does not set the array values to zero, and may therefore be marginally faster
So we are using it for faster buffer initialization